pkg/nsx/services/staticroute/staticroute.go

```diff
@@ -0,0 +1,232 @@
+package staticroute
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	common "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	util2 "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+type StaticRouteService struct {
+	common.Service
+	StaticRouteStore *StaticRouteStore
+}
+
+var (
+	log                     = logf.Log.WithName("service").WithName("staticroute")
+	resourceTypeStaticRoute = "StaticRoutes"
+	resourceType            = "resource_type"
+	String                  = common.String
+)
+
+// InitializeStaticRoute sync NSX resources
+func InitializeStaticRoute(commonService common.Service) (*StaticRouteService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+	service := &StaticRouteService{Service: commonService}
+	staticRouteStore := &StaticRouteStore{}
+	staticRouteStore.Indexer = cache.NewIndexer(keyFunc, cache.Indexers{
+		common.TagScopeStaticRouteCRUID: indexFunc,
+	})
+	staticRouteStore.BindingType = model.StaticRoutesBindingType()
+	service.StaticRouteStore = staticRouteStore
+	service.NSXConfig = commonService.NSXConfig
+	go queryStaticRoute(service, &wg, fatalErrors)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return service, err
+	}
+
+	return service, nil
+}
+
+// assume that staticroute doesn't have the same ipaddress, return true if equal
+func (service *StaticRouteService) compareStaticRoute(oldStaticRoute *model.StaticRoutes, newStaticRoute *model.StaticRoutes) bool {
+	if *oldStaticRoute.Network != *newStaticRoute.Network {
+		return false
+	}
+	oldNextHops := oldStaticRoute.NextHops
+	newNextHops := newStaticRoute.NextHops
+	if len(oldNextHops) != len(newNextHops) {
+		return false
+	}
+	oldHopsSet := sets.NewString()
+	for _, addr := range oldNextHops {
+		oldHopsSet.Insert(*addr.IpAddress)
+	}
+	for _, addr := range newNextHops {
+		if !oldHopsSet.Has(*addr.IpAddress) {
+			return false
+		}
+	}
+	return true
+}
+
+func (service *StaticRouteService) CreateorUpdateStaticRoute(obj *v1alpha1.StaticRoute) error {
+	nsxStaticRoute, err := service.buildStaticRoute(obj)
+	if err != nil {
+		return err
+	}
+
+	existingStaticRoute := service.StaticRouteStore.GetByKey(*nsxStaticRoute.Id)
+	if existingStaticRoute != nil && service.compareStaticRoute(existingStaticRoute, nsxStaticRoute) {
+		return nil
+	}
+	// TODO, update the value from 'default' to actual value
+	err = service.patch("default", "default", "default", nsxStaticRoute)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (service *StaticRouteService) buildStaticRoute(obj *v1alpha1.StaticRoute) (*model.StaticRoutes, error) {
```

Comment:
Though this file is rather small, I suggest it's better to put these build logic into a separate file builder.go to comply with our structure.